bgp: unit test speaker

ldelossa opened the pull request:
this commit adds unit tests to ensure our BGP speaker event handling
logic is error free.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgp: unit test speaker

this commit adds unit tests to ensure our BGP speaker event handling
logic is error free.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+	// and unblock our test.
+	var callCount int32
+	go func() {
+		for callCount != 3 {
```
Hmm, linter has no clue this is acting as an atomic int...

ldelossa committed:
bgp: unit test manager

this commit adds unit tests and the necessary mocks for testing our bgp
manager.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

christarazi commented on pkg/bgp/mock/gen.go:
```diff
@@ -0,0 +1,135 @@
+package mock
```
Copyright header

christarazi commented on pkg/bgp/mock/mock.go:
```diff
@@ -0,0 +1,54 @@
+// mockMetalLBSpeaker implements the speaker.Speaker interface by delegating to
+// a set of functions defined during test.
+type MockMetalLBSpeaker struct {
+	SetService_    func(name string, svc *metallbspr.Service, eps *metallbspr.Endpoints) types.SyncState
+	SetNodeLabels_ func(labels map[string]string) types.SyncState
+	PeerSession_   func() []metallbspr.Session
```
Style nit: You could simply lower-case these functions and that would be enough to disambiguate them from the interface they are a part of, instead of the not-so-obvious `_`.

ldelossa replied:
I cant because they will become private to this package and not definable from the packages under test.

christarazi replied:
Ah I didn't realize that, disregard then

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -57,7 +57,7 @@ func (s *MetalLBSpeaker) run(ctx context.Context) {
 		// which indicates the speaker should yield. shut
 		// it down.
-		if s.shutdown > 0 {
+		if s.shutdown > 0 { // atomic load not necessary, we are the only writer.
```
Are these changes meant to be part of the 1st commit?

ldelossa replied:
[1]

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -94,7 +95,7 @@ type MetalLBSpeaker struct {
 
 func (s *MetalLBSpeaker) shutDown() bool {
-	return s.shutdown > 0
+	return atomic.LoadInt32(&s.shutdown) > 0
```
Same here

ldelossa replied:
technically, these *should* have been part of the PR prior to this one, I caught this and [1] in this PR, I don't hate fixing them too much in these commits, but can re-structure it if you think that's best.

christarazi replied:
Is the idea that you're going to squash some of these in the end into the original PR's commits? If so, then I'm OK with this PR. Otherwise, I'd prefer if the history is cleaned up a bit, such that each commit is reasonably isolated.

ldelossa replied:
I dont intend to squash, and typically dont. Ill move the atomic fixes to a new commit which will become the first one in this pull.

ldelossa replied:
@christarazi I've pushed a new commit which now is the first commit of this PR which separates out and makes explicit the fixes to the atomic usages.

christarazi commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+			rr.Unlock()
+			// will unblock our test code.
+			cancel()
+			return types.SyncStateSuccess
```
While I don't think it makes a big difference, wouldn't it make more sense that `cancel()` is deferred? In other words, executing after the return. Just from a simulating the MetalLB behavior point of view. The reason is because I would imagine that the function returns first, then the test is "unblocked", but here the test is "unblocked" first, then the function returns.

ldelossa replied:
good call, I like that. Didn't think of that here.

christarazi commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+	"github.com/cilium/cilium/pkg/bgp/mock"
+	"github.com/cilium/cilium/pkg/k8s"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/google/go-cmp/cmp"
```
This is a 3rd party package, ordering :)

christarazi commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+
+// TestSpeakerOnUpdateService confirms the speaker performs the correct
+// actions when an OnUpdateService event takes place.
+func TestSpeakerOnUpdateService(t *testing.T) {
```
Comment to hold on discussion at https://cilium.slack.com/archives/C2B917YHE/p1626113446459700

christarazi commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+// actions when an OnUpdateNode event takes place.
+//
+// This test effectively covers the OnAddNode as well, as this method
+// just deleagets to OnUpdateNode with nil as the first argument to
```
Typo `deleagets`

christarazi commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+	// and unblock our test.
+	var callCount int32
+	go func() {
+		for atomic.LoadInt32(&callCount) != 3 {
```
Why 3? It would be good to state that in the comment as well

christarazi replied:
I see later down that we increment 3 times, but it wasn't obvious to me at first

christarazi commented on pkg/bgp/speaker/speaker_test.go:
```diff
@@ -0,0 +1,482 @@
+	if err := spkr.OnUpdateService(nil); err != ErrShutDown {
+		t.Fatalf("got: %v, want: %v", err, ErrShutDown)
+	}
+
```
Extra newline here

christarazi commented on pkg/bgp/mock/gen.go:
```diff
@@ -6,17 +6,20 @@ import (
 	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
 	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
 	nodetypes "github.com/cilium/cilium/pkg/node/types"
+
```
This newline should probably be part of the original commit

christarazi commented on pkg/bgp/mock/gen.go:
```diff
@@ -6,17 +6,20 @@ import (
 
 // GenTestNodeAndAdvertisements generates a slim_corev1.Node with its
-// meta details and PodCidr(s) fields hardcoded along with
+// ObjectMeta and PodCidr(s) fields hardcoded along with
```
Same here

christarazi reviewed (commented):
Awesome stuff. A few comments below which most have multiple occurrences of, not just where I commented. Also, it seems that the 2nd commit makes many changes to the 1st commit. I would expect the 2nd commit to be almost purely additive.

ldelossa committed:
bgp: fix atomic usage

previous to this commit an atomic read was performed without the
necessary atomic.Load operation.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

christarazi commented on pkg/bgp/mock/mock.go:
```diff
@@ -0,0 +1,130 @@
+	v1 "k8s.io/api/core/v1"
+)
+
+// mockMetalLBSpeaker implements the speaker.Speaker interface by delegating to
```
```suggestion
// MockMetalLBSpeaker implements the speaker.Speaker interface by delegating to
```

christarazi commented on pkg/bgp/manager/manager_test.go:
```diff
@@ -0,0 +1,173 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package manager
```
There should be a newline separate the `package ...` line and the copyright header. The reason is because the godocs will pick the copyright header as the godoc for the package.

christarazi reviewed (commented):
A few more nits. Also, I would prefer that the copyright header changes are made in the original commit that the file was introduced in. I don't think having a separate commit to add copyright headers to files introduced in the PR already as having any value. Would you mind making that change?

christarazi reviewed (approved):
Nice work!
